### ISSUE TITLE
Improve the search indexer

### ISSUE THREAD

**Issue by jonasraoni** [pkp/pkp-lib#8915]
**Describe the bug**
The current search indexer has some problems:
- It's ingesting unexpected UTF8 characters (e.g. ＆▽┼”®‒)
- The keyword insertion might fail sometimes
- It's working too slow

**What application are you using?**
OJS 3.4.0rc2

**PRs**
`stable-3_4_0`
- https://github.com/pkp/pkp-lib/pull/9215
- https://github.com/pkp/ojs/pull/3999
- https://github.com/pkp/omp/pull/1476
- https://github.com/pkp/ops/pull/585

`main`
- https://github.com/pkp/pkp-lib/pull/8916
- https://github.com/pkp/ojs/pull/3875
- https://github.com/pkp/ops/pull/588
- https://github.com/pkp/omp/pull/1478

**Comment by asmecher** [comment]
We're essentially in a testing freeze for 3.4.0-0, so I've moved this off to 3.4.0-1. Are the UTF8 characters coming from PDFs? There is some suggested filtering in the configuration file via external tools:
```
index[application/pdf] = "/usr/bin/pstotext -enc UTF-8 -nopgbrk %s - | /usr/bin/tr '[:cntrl:]' ' '"
```

**Comment by jonasraoni** [comment]
Yeah, better to merge this one after creating the 3.4.0 branch 😁

I'm not sure from where the content came, I was just taking a look at the keywords table of some installations.

As we're already filtering characters internally, I thought it was better to improve it. The `unicode escapes` worked fine for me (I've mixed a couple of different characters: punctuations, Chinese, Arabian, etc.).

**Comment by jonasraoni** [comment]
I'm not going to touch the PRs anymore. After reviewing, it will be needed to port the changes to OPS/OMP.

Some updates might be backported later if this issue is still pertinent: https://github.com/pkp/pkp-lib/issues/7193.

**Comment by asmecher** [comment]
@jonasraoni, this was previously scheduled for 3.4.0-x, but the PRs are against `main` and I note that we previously talked about merging after `stable-3_4_0` was forked from `main`. Should this actually be filed against 3.5.0?

**Comment by jonasraoni** [comment]
@asmecher I think it's ok to get it merged in 3.4.0-x, but to fix it properly, a full content reindex will be needed, due to the existing bad keywords damaging the search 🤔

If you agree with the reindexing, then it should probably be moved to 3.5. 

Another option would be to update the reindexer to work in batches (it should be done anyway). Then in my opinion it would be ok to have this extra step on the migration, as it won't compromise the minor upgrade speed.

**Comment by asmecher** [comment]
@jonasraoni, thanks -- these look good. But they'll need to be ported to OMP and OPS too, or else `stable-3_4_0` will be broken for those apps, right?

**Comment by asmecher** [comment]
(I'm open to putting this back into 3.4.0-2 if it can be done before I wrap up the release!)

**Comment by jonasraoni** [comment]
Ok! As the 3.4.0-2 will be delayed by a couple of days, I think I'll have enough time to create the PRs, and do a general test using the external indexers.

**Comment by jonasraoni** [comment]
@asmecher I've created the remaining PRs. Just the PR for OMP had extra additions (I've added missing tests and updated the code to look like in the other applications).

**Comment by asmecher** [comment]
@jonasraoni, it looks like there are a few PRs scattered to various repos and branches for this one; can you summarize the relevant PRs in the issue description?

**Comment by jonasraoni** [comment]
@asmecher I've added the PRs to the description. Notice the PRs for OMP have slightly extra updates due to differences that I found regarding the other applications.

**Comment by asmecher** [comment]
All merged -- thanks, @jonasraoni!
